include/net/net_pkt.h

```diff
@@ -58,8 +58,18 @@ struct net_pkt_cursor {
  * net_pkt_clone() function.
  */
 struct net_pkt {
-	/** Internal variable that is used when packet is sent */
-	struct k_work work;
+	union {
+		/** Internal variable that is used when packet is sent
+		 * or received.
+		 */
+		struct k_work work;
+		/** Socket layer will queue received net_pkt into a k_fifo.
+		 * Since this happens after consuming net_pkt's k_work on
+		 * RX path, it is then fine to have both attributes sharing
+		 * the same memory area.
+		 */
+		int sock_recv_fifo;
```

Comment:
Actually, probably worth thinking about 64-bit cleanness. So, ping to @andyross and @andrewboie who might have been involved with x86_64 port re: whether it's better to have that as `void *` instead. It was `int` before https://github.com/zephyrproject-rtos/zephyr/commit/79672d1647d9d005cd295cb63b10f801ca97dea2 though, which this PR fixes, so keeping it as `int` won't make anything worse than it was.